examples/src/main/java/com/datastax/oss/driver/examples/concurrent/LimitConcurrencyCustomAsync.java

```diff
@@ -0,0 +1,191 @@
+import static com.datastax.oss.driver.api.querybuilder.QueryBuilder.insertInto;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.CqlSessionBuilder;
+import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+
+/**
+ * Creates a keyspace and tables, and loads data using Async API into them.
+ *
+ * <p>This example makes usage of a {@link CqlSession#executeAsync(String)} method, which is
+ * responsible for executing requests in a non-blocking way. It uses {@link CompletableFuture} to
+ * limit number of concurrent request to {@code CONCURRENCY_LEVEL}.
+ *
+ * <p>Preconditions:
+ *
+ * <ul>
+ *   <li>An Apache Cassandra(R) cluster is running and accessible through the contacts points
+ *       identified by basic.contact-points (see application.conf).
+ * </ul>
+ *
+ * <p>Side effects:
+ *
+ * <ul>
+ *   <li>creates a new keyspace "examples" in the session. If a keyspace with this name already
+ *       exists, it will be reused;
+ *   <li>creates a table "examples.tbl_sample_kv". If it exist already, it will be reused;
+ *   <li>inserts a TOTAL_NUMBER_OF_INSERTS of rows into the table.
+ * </ul>
+ *
+ * @see <a href="http://datastax.github.io/java-driver/manual/">Java driver online manual</a>
+ */
+public class LimitConcurrencyCustomAsync {
+  private static final int CONCURRENCY_LEVEL = 32;
+  private static final int TOTAL_NUMBER_OF_INSERTS = 10_000;
+  // Used to track number of total inserts
+  private static final AtomicInteger INSERTS_COUNTER = new AtomicInteger();
+
+  public static void main(String[] args) throws InterruptedException, ExecutionException {
+
+    try (CqlSession session = new CqlSessionBuilder().build()) {
+      createSchema(session);
+      insertConcurrent(session);
+    }
+  }
+
+  private static void insertConcurrent(CqlSession session)
+      throws InterruptedException, ExecutionException {
+    PreparedStatement pst =
+        session.prepare(
+            insertInto("examples", "tbl_sample_kv")
+                .value("id", bindMarker("id"))
+                .value("value", bindMarker("value"))
+                .build());
+
+    // Construct CONCURRENCY_LEVEL number of ranges.
+    // Each range will be executed independently.
+    List<Range> ranges = createRanges(CONCURRENCY_LEVEL, TOTAL_NUMBER_OF_INSERTS);
+
+    // List of pending CONCURRENCY_LEVEL features that we will wait for at the end of the program.
+    List<CompletableFuture<?>> pending = new ArrayList<>();
+
+    // Every range will have dedicated CompletableFuture handling the execution.
+    for (Range range : ranges) {
```

Comment:
This is an interesting way of controlling concurrency but it suffers a weakness in that if you have a slow response it may delay submitting requests in that range while other ranges may have otherwise completed, so you aren't effectively optimizing your concurrency.  For the intent of the example it's fine though, especially since it mirrors the nodejs implementation.

An alternative way to accomplish this would be to submit requests from the same thread, and on completion notify/wake up the thread in some way (could be as simple as enqueuing the next task into a queue) to evaluate the current number of inflight requests and then submit more.  I think we don't need to go that far though.

Reply:
Or better yet:

```
private static void insertConcurrent(CqlSession session) throws InterruptedException {
  PreparedStatement pst =
      session.prepare(
          insertInto("examples", "tbl_sample_kv")
              .value("id", bindMarker("id"))
              .value("value", bindMarker("value"))
              .build());

  // Executor service with CONCURRENCY_LEVEL number of threads that states an upper limit
  // on number of request in progress.
  ExecutorService executor = Executors.newFixedThreadPool(CONCURRENCY_LEVEL);

  // create statements
  Queue<BoundStatement> statements = new ConcurrentLinkedQueue<>();
  for (int i = 0; i < TOTAL_NUMBER_OF_INSERTS; i++) {
    statements.add(pst.bind().setUuid("id", UUID.randomUUID()).setInt("value", i));
  }

  // execute statements
  for (int i = 0; i < CONCURRENCY_LEVEL; i++) {
    executor.submit(() -> acquireQueryAndRelease(executor, session, statements));
  }

  REQUEST_LATCH.await();

  // Shutdown executor to free resources
  executor.shutdown();
}

private static void acquireQueryAndRelease(
    ExecutorService executor, CqlSession session, Queue<BoundStatement> statements) {
  BoundStatement stmt = statements.poll();
  if (stmt == null) {
    return;
  }
  SEMAPHORE.acquireUninterruptibly();
  session
      .executeAsync(stmt)
      .whenCompleteAsync(
          (result, error) -> {
            REQUEST_LATCH.countDown();
            SEMAPHORE.release();
            if (error != null) {
              error.printStackTrace();
            }
            acquireQueryAndRelease(executor, session, statements);
          },
          executor);
}
```

Reply:
That's a good solution :+1: and I like that better than my suggestion.  One suggestion, since the generated statements are based on an incrementing value, is there a need to pregenerate them?  You could just use an atomic long and increment and generate the statements as you execute them, and then return in `acquireQueryAndRelease` when the long is above `TOTAL_NUMBER_OF_INSERTS`.

Reply:
I think for the purpose of this example, having multiple sequential async executions is good enough. It's true that it may suffer from a single slow response in each `executeOneAtATime()` call, but with a large enough sample it should average out.

I would like to avoid generating the statements upfront in a queue as we will be allocating all the memory upfront, making the example unsuitable for unbounded `TOTAL_NUMBER_OF_INSERTS` values.

Reply:
Unbounded inserts are rare. I still believe that 99% of users will have if not the statements, at least the data ready in memory before the insert operation begins. Large datasets will typically be broken into smaller batches of a few 1000 rows if need be. But in any case, nobody generates data to be inserted on-the-fly, unless you are populating the database with random data.
Truly unbounded use cases would imo be typically handled by using the java Stream API, a Producer-Consumer setup with shared queue, or by reactive programming.

Reply:
I think that the main question is: do we want to have parity between examples in nodejs and java. Currently, the nodejs example is doing it using ranges and existing java example does the same. I like @adutra example because is more java idiomatic but still there is a problem of throttling requests - if we will have unbounded `Queue` there could be an out of memory error.
This line: 
```
  for (int i = 0; i < TOTAL_NUMBER_OF_INSERTS; i++) {
    statements.add(pst.bind().setUuid("id", UUID.randomUUID()).setInt("value", i));
  }
```
still is not solving that problem of out of memory on statements.

Reply:
1. I think it's better if each driver strives to present idiomatic examples for its language.
2. About the potential out of memory problem: this is just an example, so why not simply add a comment above the for-loop that creates the statements:

> The code below creates statements in a loop for the sake of simplicity. In real life, you would probably obtain the data to insert from an external resource or system (e.g. a file, or a REST service); care should be taken not to run out of memory if too much data is held in memory.

Reply:
@adutra Looking at your example I think that it is almost the same as existing: `LimitConcurrencyCustom`. Both are using `SEMAPHORE` to limit `IN_FLIGHT` and both are using     
`ExecutorService executor = Executors.newFixedThreadPool(CONCURRENCY_LEVEL);` to limit number of concurrent requests. There is only one difference - in your example, you are using intermediate `Queue`.
My suggestion is like this:
- I can add a `Queue` to the `LimitConcurrencyCustom` example to make a clearer distinction between consumer and producer - this will be java idiomatic way OR add the 3rd example with a Queue.
- keep `LimitConcurrencyCustomAsync` according to @jorgebay suggestion to also have an example that is showing the throttling of produced queries (and is one to one with `NodeJs` example).
Wdyt?

Reply:
I agree with @tomekl007 about keeping `LimitConcurrencyCustomAsync` in the current state, as it shows a technique that chains futures without relying on any external component and concepts (latch / queue / semaphore). It covers important concepts that are hard to grasp for most users in few lines of code.

Reply:
Ok to keep the code as is. I don't mind, really.